p2p/starknet/handlers.go

```diff
@@ -112,75 +120,64 @@ func (h *Handler) CurrentBlockHeaderHandler(stream network.Stream) {
-	if err != nil {
-		return nil, err
-	}
-	return h.blockHeaders(it, blockHeadersRequestFin()), nil
+	return h.onBlockHeadersRequest(&spec.BlockHeadersRequest{
+		Iteration: &spec.Iteration{
+			Start: &spec.Iteration_BlockNumber{
+				BlockNumber: curHeight,
+			},
+			Direction: spec.Iteration_Forward,
+			Limit:     1,
+			Step:      1,
+		},
+	})
 }
 
 func (h *Handler) onBlockHeadersRequest(req *spec.BlockHeadersRequest) (iter.Seq[proto.Message], error) {
-	it, err := h.newIterator(req.Iteration)
-	if err != nil {
-		return nil, err
+	finMsg := &spec.BlockHeadersResponse{
+		Part: []*spec.BlockHeadersResponsePart{
+			{
+				HeaderMessage: &spec.BlockHeadersResponsePart_Fin{},
+			},
+		},
 	}
-	return h.blockHeaders(it, blockHeadersRequestFin()), nil
-}
 
-func (h *Handler) blockHeaders(it *iterator, fin Stream[proto.Message]) iter.Seq[proto.Message] {
-	return func(yield func(proto.Message) bool) {
-		for it.Valid() {
-			header, err := it.Header()
-			if err != nil {
-				h.log.Debugw("Failed to fetch header", "blockNumber", it.BlockNumber(), "err", err)
-				break
-			}
+	return h.processIterationRequest(req.Iteration, finMsg, func(it blockDataAccessor) (proto.Message, error) {
+		header, err := it.Header()
+		if err != nil {
+			return nil, err
+		}
 
-			h.log.Debugw("Created Header Iterator", "blockNumber", header.Number)
+		h.log.Debugw("Created Header Iterator", "blockNumber", header.Number)
 
-			commitments, err := h.bcReader.BlockCommitmentsByNumber(header.Number)
-			if err != nil {
-				h.log.Debugw("Failed to fetch block commitments", "blockNumber", it.BlockNumber(), "err", err)
-				break
-			}
+		commitments, err := h.bcReader.BlockCommitmentsByNumber(header.Number)
+		if err != nil {
+			return nil, err
+		}
 
-			msg := &spec.BlockHeadersResponse{
-				Part: []*spec.BlockHeadersResponsePart{
-					{
-						HeaderMessage: &spec.BlockHeadersResponsePart_Header{
-							Header: core2p2p.AdaptHeader(header, commitments),
-						},
+		return &spec.BlockHeadersResponse{
+			Part: []*spec.BlockHeadersResponsePart{
+				{
+					HeaderMessage: &spec.BlockHeadersResponsePart_Header{
+						Header: core2p2p.AdaptHeader(header, commitments),
 					},
-					{
-						HeaderMessage: &spec.BlockHeadersResponsePart_Signatures{
-							Signatures: &spec.Signatures{
-								Block:      core2p2p.AdaptBlockID(header),
-								Signatures: utils.Map(header.Signatures, core2p2p.AdaptSignature),
-							},
+				},
+				{
+					HeaderMessage: &spec.BlockHeadersResponsePart_Signatures{
+						Signatures: &spec.Signatures{
+							Block:      core2p2p.AdaptBlockID(header),
+							Signatures: utils.Map(header.Signatures, core2p2p.AdaptSignature),
 						},
 					},
 				},
-			}
-
-			if !yield(msg) {
-				return
-			}
-
-			it.Next()
-		}
-
-		if finMsg, ok := fin(); ok {
-			yield(finMsg)
-		}
-	}
+			},
+		}, nil
+	})
 }
 
 func (h *Handler) onBlockBodiesRequest(req *spec.BlockBodiesRequest) (iter.Seq[proto.Message], error) {
```

Comment:
Can we use processIterationRequest here?

Reply:
unfortunately now, this handler is different because it sends multiple different messages for 1 request. But spoiler: in updated spec this method is going to be removed.